src/common/adapters/MultiAdapter.sol

```diff
@@ -0,0 +1,226 @@
+            } else {
+                gateway.handle(centrifugeId, payload);
+            }
+
+            // Only if there are no more pending messages, remove the pending message
+            if (state.votes.isEmpty()) {
+                delete state.pendingBatch;
+            }
+        } else if (!isMessageProof) {
+            state.pendingBatch = payload;
+        }
+    }
+
+    function initiateRecovery(uint16 centrifugeId, IAdapter adapter, bytes32 payloadHash) external auth {
+        require(_activeAdapters[centrifugeId][adapter].id != 0, InvalidAdapter());
+        recoveries[centrifugeId][adapter][payloadHash] = block.timestamp + RECOVERY_CHALLENGE_PERIOD;
+        emit InitiateRecovery(centrifugeId, payloadHash, adapter);
+    }
+
+    function disputeRecovery(uint16 centrifugeId, IAdapter adapter, bytes32 payloadHash) external auth {
+        delete recoveries[centrifugeId][adapter][payloadHash];
+        emit DisputeRecovery(centrifugeId, payloadHash, adapter);
+    }
+
+    function executeRecovery(uint16 centrifugeId, IAdapter adapter, bytes calldata payload) external {
+        bytes32 payloadHash = keccak256(payload);
+        uint256 recovery = recoveries[centrifugeId][adapter][payloadHash];
+
+        require(recovery != 0, RecoveryNotInitiated());
+        require(recovery <= block.timestamp, RecoveryChallengePeriodNotEnded());
+
+        delete recoveries[centrifugeId][adapter][payloadHash];
+        _handle(centrifugeId, payload, adapter);
+        emit ExecuteRecovery(centrifugeId, payload, adapter);
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Outgoing
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IAdapter
+    function send(uint16 centrifugeId, bytes calldata payload, uint256 gasLimit, address refund)
+        external
+        payable
+        returns (bytes32)
+    {
+        require(msg.sender == address(gateway), NotGateway());
+
+        IAdapter[] memory adapters_ = adapters[centrifugeId];
+        bytes32 hash = keccak256(payload);
+        bytes32 payloadId = keccak256(abi.encodePacked(localCentrifugeId, centrifugeId, hash));
+        bytes memory proof = hash.serializeMessageProof();
+
+        for (uint256 i; i < adapters_.length; i++) {
+            bytes memory adapterPayload = i == PRIMARY_ADAPTER_ID - 1 ? payload : proof;
+
+            uint256 cost = adapters_[i].estimate(centrifugeId, adapterPayload, gasLimit);
+            bytes32 adapterData = adapters_[i].send{value: cost}(centrifugeId, adapterPayload, gasLimit, refund);
+
+            if (i == PRIMARY_ADAPTER_ID - 1) {
+                emit SendBatch(centrifugeId, payloadId, payload, adapters_[i], adapterData, refund);
+            } else {
+                emit SendProof(centrifugeId, payloadId, hash, adapters_[i], adapterData);
+            }
+        }
+
+        return bytes32(0);
```

Comment:
Nit Q: Can we drop this return as it always returns the same empty zero bytes32?

Reply:
The problem here is that it implements a method defined in `IAdapter` that returns that value.

@hieronx could you extend a bit more about `adapterData`? Do we really need it? It's only used in one event, and only wormhole supports that.

Reply:
> could you extend a bit more about adapterData? Do we really need it? It's only used in one event, and only wormhole supports that.

It will be really useful for linking gateway events to wormhole events though. I would expect more adapters to support this, but even if not, I would say it's already worth it for wormhole.